### PR TITLE
[API]adding promotions with expire date

### PR DESCRIPTION
**Pull request by arti0090**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/spec/CommandHandler/Cart/ApplyCouponToCartHandlerSpec.php [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+        $promotionCouponRepository->findOneBy(['code' => 'couponCode'])->willReturn($promotion);
+
+        $promotion->getUsageLimit()->willReturn(1);
+
+        $promotion->getExpiresAt()->willReturn(new \DateTime('now -1 year'));
+
+        $this->shouldThrow(\InvalidArgumentException::class)->during('__invoke', [$command]);
+    }
+
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/spec/CommandHandler/Cart/ApplyCouponToCartHandlerSpec.php:118]:
```suggestion
```

File: src/Sylius/Bundle/ApiBundle/spec/CommandHandler/Cart/ApplyCouponToCartHandlerSpec.php [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace spec\Sylius\Bundle\ApiBundle\CommandHandler\Cart;
+
+use _HumbugBox89320708a2e3\Nette\InvalidArgumentException;
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/spec/CommandHandler/Cart/ApplyCouponToCartHandlerSpec.php:16]: ?

File: src/Sylius/Behat/Context/Api/Shop/PromotionContext.php [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+use Sylius\Behat\Service\SharedStorage;
+use Symfony\Component\HttpFoundation\Request as HttpRequest;
+use Webmozart\Assert\Assert;
+
+final class PromotionContext implements Context
+{
+    /** @var ApiClientInterface */
+    private $cartsClient;
+
+    /** @var SharedStorage */
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/PromotionContext.php:29]:
```suggestion
    /** @var SharedStorageInterface */
```
and this service could be placed as the last one

File: src/Sylius/Behat/Context/Api/Shop/PromotionContext.php [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    private $sharedStorage;
+
+    /** @var ResponseCheckerInterface */
+    private $responseChecker;
+
+    public function __construct(ApiClientInterface $cartsClient, SharedStorage $sharedStorage, ResponseCheckerInterface $responseChecker)
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/PromotionContext.php:35]: As above

File: src/Sylius/Behat/Context/Api/Shop/PromotionContext.php [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        $this->useCouponCode($couponCode);
+    }
+
+    /**
+     * @Then I should be notified that the coupon is invalid
+     */
+    public function iShouldBeNotifiedThatCouponIsInvalid(): void
+    {
+        Assert::same($this->cartsClient->getLastResponse()->getStatusCode(), 500);
+        Assert::notNull($this->responseChecker->getError($this->cartsClient->getLastResponse()));
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/PromotionContext.php:56]:
```suggestion
        $response = $this->cartsClient->getLastResponse();

        Assert::same($response->getStatusCode(), 500);
        Assert::notNull($this->responseChecker->getError($response));
```

File: src/Sylius/Bundle/ApiBundle/CommandHandler/Cart/ApplyCouponToCartHandler.php [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+final class ApplyCouponToCartHandler implements MessageHandlerInterface
+{
+    /** @var OrderRepositoryInterface $orderRepository */
+    private $orderRepository;
+
+    /** @var PromotionCouponRepositoryInterface $promotionCouponRepository */
+    private $promotionCouponRepository;
+
+    /** @var OrderProcessorInterface $orderProcessor */
+    private $orderProcessor;
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/Cart/ApplyCouponToCartHandler.php:36]:
```suggestion
    /** @var OrderRepositoryInterface */
    private $orderRepository;

    /** @var PromotionCouponRepositoryInterface */
    private $promotionCouponRepository;

    /** @var OrderProcessorInterface */
    private $orderProcessor;
```

File: src/Sylius/Bundle/ApiBundle/spec/CommandHandler/Cart/ApplyCouponToCartHandlerSpec.php [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        PromotionCouponInterface $promotion,
+        OrderProcessorInterface $orderProcessor
+    ): void {
+        $orderRepository->findCartByTokenValue('cart')->willReturn($cart);
+
+        $promotionCouponRepository->findOneBy(['code' => 'couponCode'])->willReturn($promotion);
+
+        $promotion->getUsageLimit()->willReturn(null);
+
+        $promotion->getExpiresAt()->willReturn(null);
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/CommandHandler/Cart/ApplyCouponToCartHandlerSpec.php:53]:
```suggestion
        $promotionCouponRepository->findOneBy(['code' => 'couponCode'])->willReturn($promotion);
        $promotion->getUsageLimit()->willReturn(null);
        $promotion->getExpiresAt()->willReturn(null);
```

File: src/Sylius/Bundle/ApiBundle/spec/CommandHandler/Cart/ApplyCouponToCartHandlerSpec.php [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    {
+        $this->shouldImplement(MessageHandlerInterface::class);
+    }
+
+    function it_applies_coupon_to_cart(
+        OrderRepositoryInterface $orderRepository,
+        OrderInterface $cart,
+        PromotionCouponRepositoryInterface $promotionCouponRepository,
+        PromotionCouponInterface $promotion,
+        OrderProcessorInterface $orderProcessor
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/CommandHandler/Cart/ApplyCouponToCartHandlerSpec.php:45]: Arguments from `let` method should be placed as the first one in specs
```suggestion
        OrderRepositoryInterface $orderRepository,
        PromotionCouponRepositoryInterface $promotionCouponRepository,
        OrderProcessorInterface $orderProcessor,
        OrderInterface $cart,
        PromotionCouponInterface $promotion
```

File: src/Sylius/Bundle/ApiBundle/spec/CommandHandler/Cart/ApplyCouponToCartHandlerSpec.php [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+        $command = new ApplyCouponToCart('couponCode');
+        $command->setOrderTokenValue('cart');
+
+        $orderRepository->findCartByTokenValue('cart')->willReturn($cart);
+
+        $promotionCouponRepository->findOneBy(['code' => 'couponCode'])->willReturn($promotion);
+
+        $promotion->getUsageLimit()->willReturn(1);
+
+        $promotion->getExpiresAt()->willReturn(new \DateTime('now -1 year'));
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/CommandHandler/Cart/ApplyCouponToCartHandlerSpec.php:110]:
```suggestion
        $promotionCouponRepository->findOneBy(['code' => 'couponCode'])->willReturn($promotion);
        $promotion->getUsageLimit()->willReturn(1);
        $promotion->getExpiresAt()->willReturn(new \DateTime('now -1 year'));
```

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thank you, @arti0090! :tada:

File: src/Sylius/Behat/Context/Api/Shop/PromotionContext.php [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    /**
+     * @Then I should be notified that the coupon is invalid
+     */
+    public function iShouldBeNotifiedThatCouponIsInvalid(): void
+    {
+        $response = $this->cartsClient->getLastResponse();
+
+        Assert::same($response->getStatusCode(), 500);
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Shop/PromotionContext.php:60]: 500 is a wrong indicator of wrong coupon code usage. We should add validation to this feature.

File: src/Sylius/Bundle/ApiBundle/CommandHandler/Cart/ApplyCouponToCartHandler.php [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+        $promotion = $promotionCoupon->getPromotion();
+
+        Assert::notNull($promotion, 'Could not find promotion linked with this coupon');
+
+        $promotionEndDate = $promotion->getEndsAt();
+
+        if (null !== $promotionEndDate) {
+            Assert::greaterThan($promotionEndDate, new DateTime(), 'Promotion has expired');
+        }
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/Cart/ApplyCouponToCartHandler.php:80]: Do we have to check all these things to perform this action? Most of it seems to be related to validation and processing logic.

**Pull request merged**